### PR TITLE
add analytic true impes

### PR DESCRIPTION
**Pull request by hnil**
-- give analytic cpr weights
-- useful for thermal
-- useful for cases with strange pvt

### CODE REVIEW

File: opm/simulators/linalg/ISTLSolverEbos.hpp [hunk 1]
```diff
@@ -459,16 +459,11 @@ std::unique_ptr<Matrix> blockJacobiAdjacency(const Grid& grid,
...
-                    [this]
-                    {
-                        return this->getTrueImpesWeights(pressureIndex);
-                    };
-
                 if (!useWellConn_) {
                     auto wellOp = std::make_unique<WellModelOperator>(simulator_.problem().wellModel());
                     flexibleSolver_[activeSolverNum_].wellOperator_ = std::move(wellOp);
                 }
+                std::function<Vector()> trueFunc = this->getWeightsCalculator(prm_[activeSolverNum_], getMatrix(), pressureIndex);
```
[the remark this answers is not in the document]
Comment (hnil) [reply opm/simulators/linalg/ISTLSolverEbos.hpp:466]: changed.

File: opm/simulators/linalg/getQuasiImpesWeights.hpp [hunk 2]
```diff
@@ -126,12 +126,118 @@ namespace Amg
...
+        using PrimaryVariables = typename Model::PrimaryVariables;
+        using VectorBlockType = typename Vector::block_type;
+        constexpr int numEq = VectorBlockType::size();
+        using Evaluation =
+            typename std::decay_t<decltype(model.localLinearizer(threadId).localResidual().residual(0))>::block_type;
+        using Toolbox = MathToolbox<Evaluation>;
+        VectorBlockType rhs(0.0);
+        const auto& solution = model.solution(/*timeIdx*/ 0);
+        OPM_BEGIN_PARALLEL_TRY_CATCH();
+         for (const auto& elem : elements(gridView)) {
```
[the remark this answers is not in the document]
Comment (hnil) [reply opm/simulators/linalg/getQuasiImpesWeights.hpp:165]: fixed

File: opm/simulators/linalg/getQuasiImpesWeights.hpp [hunk 3]
```diff
@@ -126,12 +126,118 @@ namespace Amg
...
+            if (priVars.primaryVarsMeaningGas() == PrimaryVariables::GasMeaning::Rs) {
+                rv = 0.0;
+            }
+            if (FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx)
+                && FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {
+                denominator = Toolbox::template decay<LhsEval>(1 - rs * rv);
+            }
+
+            if (not(denominator > 0)) {
+                std::cout << "Probably negative compressibility" << std::endl;
```
[the remark this answers is not in the document]
Comment (hnil) [reply opm/simulators/linalg/getQuasiImpesWeights.hpp:196]: Just deleted.

File: opm/simulators/linalg/getQuasiImpesWeights.hpp [hunk 4]
```diff
@@ -126,12 +126,118 @@ namespace Amg
...
+                    / denominator);
+            }
+            if (FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {
+                unsigned activeCompIdx = Indices::canonicalToActiveComponentIndex(
+                    FluidSystem::solventComponentIndex(FluidSystem::gasPhaseIdx));
+                bweights[activeCompIdx] = Toolbox::template decay<LhsEval>(
+                    (1 / fs.invB(FluidSystem::gasPhaseIdx) - rv / fs.invB(FluidSystem::oilPhaseIdx))
+                    / denominator);
+            }
+            if(false){
```
[the remark this answers is not in the document]
Comment (hnil) [reply opm/simulators/linalg/getQuasiImpesWeights.hpp:212]: Just deleted.

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
All green, merging!

**Comment by blattms** [comment]
There seems to be a new parameter value, right? I think we should add it to the help message in a separate PR.

**Pull request merged**